### PR TITLE
Feature/distribute delegate

### PR DESCRIPTION
**Pull request by jaelim-ms**

### CODE REVIEW

**Comment by jaelim-ms** [comment]
Please do not merge this PR. This is not a part of next release.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 1]
```diff
@@ -128,8 +119,58 @@ + (void)openUrl:(NSURL *)url {
...
+}
+
++ (void)setDelegate:(id<MSDistributeDelegate>)delegate {
+  [[self sharedInstance] setDelegate:delegate];
+}
+
 #pragma mark - Private
 
+- (void)startUpdate {
+  if ([self isEnabled]) {
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Isn't this check redundant?

File: MobileCenterDistribute/MobileCenterDistribute/MSDistributeDelegate.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+@optional
+
+/**
+ * Callback method that will be called whenever a new release is available for update.
+ *
+ * @param releaseDetails Release details for the update.
+ *
+ * @return Return YES and if you want to take update control by overriding default update dialog, NO otherwise.
+ *
+ * @seealso [MSDistribute notifyUserUpdateAction:]
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistributeDelegate.h]: Does `@seealso` one really exists or should we use `@see` instead?

File: MobileCenterDistribute/MobileCenterDistribute/MSDistributeDelegate.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+@optional
+
+/**
+ * Callback method that will be called whenever a new release is available for update.
+ *
+ * @param releaseDetails Release details for the update.
+ *
+ * @return Return YES and if you want to take update control by overriding default update dialog, NO otherwise.
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistributeDelegate.h]: Maybe the `and` can be removed.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistributePrivate.h [hunk 4]
```diff
@@ -74,9 +69,14 @@ static NSString *const kMSUpdateTokenKey = @"MSUpdateToken";
 @property(nonatomic) MSAlertController *updateAlertController;
 
 /**
- * Current mandatory release if any.
+ * Current release details.
+ */
+@property(nonatomic) MSReleaseDetails *releaseDetails;
+
+/**
+ * A Distribute delegate that will be called whenever a new release is available for update
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistributePrivate.h]: `.` missing at the end.

File: MobileCenterDistribute/MobileCenterDistribute/Resources/fr.lproj/MobileCenterDistribute.strings [hunk 5]
```diff
@@ -1,6 +1,5 @@
-"Ignore" = "Ignorer";
 "Postpone" = "Reporter";
-"Download" = "Télécharger";
+"Update" = "Update";
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/Resources/fr.lproj/MobileCenterDistribute.strings]: "Mettre à jour" if we want to translate :)

File: MobileCenterDistribute/MobileCenterDistributeTests/MSDistributeTests.m [hunk 6]
```diff
@@ -1072,6 +1067,140 @@ - (void)testIsNewerVersionFunction {
...
+    [showConfirmationAlertExpectation fulfill];
+  });
+
+  // Then
+  [self waitForExpectationsWithTimeout:1
+                               handler:^(NSError *error) {
+                                 if (error) {
+                                   XCTFail(@"Expectation Failed with error: %@", error);
+                                 }
+                                 OCMReject([distributeMock showConfirmationAlert:detailsMock]);
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistributeTests/MSDistributeTests.m]: Should verify that `onNewUpdateAvailable` has been called on the delegate.

File: MobileCenterDistribute/MobileCenterDistributeTests/MSDistributeTests.m [hunk 7]
```diff
@@ -1072,6 +1067,140 @@ - (void)testIsNewerVersionFunction {
...
+    [showConfirmationAlertExpectation fulfill];
+  });
+
+  // Then
+  [self waitForExpectationsWithTimeout:1
+                               handler:^(NSError *error) {
+                                 if (error) {
+                                   XCTFail(@"Expectation Failed with error: %@", error);
+                                 }
+                                 OCMVerify([distributeMock showConfirmationAlert:detailsMock]);
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistributeTests/MSDistributeTests.m]: Should verify that `onNewUpdateAvailable` has been called on the delegate.

File: Puppet/Puppet/AppDelegate.m [hunk 8]
```diff
@@ -149,4 +149,33 @@ - (void)crashes:(MSCrashes *)crashes didFailSendingErrorReport:(MSErrorReport *)
   NSLog(@"Did fail sending report with: %@, and error: %@", errorReport.exceptionReason, error.localizedDescription);
 }
 
+#pragma mark - MSDistributeDelegate
+
+- (BOOL)onNewUpdateAvailable:(MSReleaseDetails *)releaseDetails {
```
Comment (clpolet) [inline Puppet/Puppet/AppDelegate.m]: We should review the naming for this callback.

File: Puppet/Puppet/Constants.h [hunk 9]
```diff
@@ -2,4 +2,7 @@
  * Copyright (c) Microsoft Corporation. All rights reserved.
  */
 
+#import <Foundation/Foundation.h>
+
 static NSString *const kPUPLogTag = @"[Puppet]";
+static NSString *const kMSCustomizedUpdateAlertKey = @"kMSCustomizedUpdateAlertKey";
```
Comment (clpolet) [inline Puppet/Puppet/Constants.h]: That's the puppet so I would suggest not using the `MS` prefix here to better fit a real application.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m (hunk 1 above)
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Well, that cannot be if it is called from another place. BTW, I found this has a bug and some of changes made in here will be introduced in Update Dialog change PR. I want to close it temporary and then reopen once it is ready for merge.

**Comment by jaelim-ms** [comment]
Closing the PR due to release.

File: Puppet/Puppet/AppDelegate.m (hunk 8 above)
Comment (jaelim-ms) [inline Puppet/Puppet/AppDelegate.m]: This is as of now. Still waiting for the final name from PM.

**Comment by jaelim-ms** [comment]
Instead of resolving the conflicts, close this one. #492 is a new PR.

**Pull request closed without merging**